### PR TITLE
add block_mask & retrain_free features

### PR DESCRIPTION
**Pull request by WeiweiZhang1**
## Type of Change
Pruning Features and Bugfix

## Details
add block_mask & retrain_free pruning-types
fixed the pattern Initialization problem and the problem of substandard sparsity.
add feature related LLM model pruning and slim example and test-units.

API not changed

### CODE REVIEW

File: neural_compressor/compression/pruner/criteria.py [hunk 1]
```diff
@@ -190,3 +190,89 @@ def on_before_optimizer_step(self):
...
+            self.scores[key] = torch.zeros(mask.shape).to(mask.device)
+            self.collected_grads[key] = []
+
+    def on_before_optimizer_step(self):
+        """Calculate and store the pruning scores based on snip_momentum_block criterion."""
+        with torch.no_grad():
+            for key in self.modules.keys():
+                if not hasattr(self.modules[key], 'block_mask'):
+                    continue # No corresponding block mask, skip.
+                mask_grad = self.modules[key].block_mask.grad.clone()
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/criteria.py:274]: don't forget to remove the clone

File: neural_compressor/compression/pruner/criteria.py [hunk 2]
```diff
@@ -190,3 +190,89 @@ def on_before_optimizer_step(self):
...
+                self.collected_grads[key].append(mask_grad)
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/criteria.py:275]: why not use +, append will cost much more memory

File: neural_compressor/compression/pruner/pruners.py [hunk 3]
```diff
@@ -514,7 +515,7 @@ def update_block_masks(self, masks):
         with torch.no_grad():
             for key in self.masks.keys():
                 module = self.modules[key]
-                module.block_mask.data = masks[key].data
+                module.block_mask.data = masks[key].data.clone()
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruners.py:518]: please review all the clone operation to check whether it is a must

File: neural_compressor/compression/pruner/criteria.py [hunk 4]
```diff
@@ -190,3 +190,89 @@ def on_before_optimizer_step(self):
...
+    def __init__(self, modules, config):
+        """Initiliaze a block_mask pruning criterion."""
+        super(SnipMomentumBlockCriterion, self).__init__(modules, config)
+        assert self.config.end_step > 0, "please set end_step > 0 for gradient based criterion"
+        for key in self.modules.keys():
+            if not hasattr(self.modules[key], 'block_mask'):
+                continue # No corresponding block mask, skip.
+            mask = self.modules[key].block_mask
+            self.scores[key] = torch.zeros(mask.shape).to(mask.device)
+        self.alpha = 1.0
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/criteria.py:221]: why set alpha=1.0

File: examples/pytorch/nlp/huggingface_models/language-modeling/pruning/eager/run_clm_no_trainer.py [hunk 5]
```diff
@@ -0,0 +1,915 @@
...
+                output_dir = os.path.join(args.output_dir, output_dir)
+            accelerator.save_state(output_dir)
+
+    compression_manager.callbacks.on_train_end()
+    
+    #-----------------------------start auto slim----------------------------------#
+    if args.auto_slim:
+        logger.info(f"***** Running Evaluation before ffn auto slim*****")
+        
+        accuracy, avg_latency = eval_func(model)
```
Comment (wenhuach21) [inline examples/pytorch/nlp/huggingface_models/language-modeling/pruning/eager/run_clm_no_trainer.py:886]: move these 3 lines out of auto_slim

File: examples/pytorch/nlp/huggingface_models/language-modeling/pruning/eager/run_llm_pruning.sh
Comment (YIYANGCAI) [inline examples/pytorch/nlp/huggingface_models/language-modeling/pruning/eager/run_llm_pruning.sh:1]: replace path with string like "/path/to/your/model" will be easier to understand.

**Review by YIYANGCAI — APPROVED** [review]

File: neural_compressor/compression/pruner/pruners.py [hunk 6]
```diff
@@ -284,6 +284,31 @@ def check_is_pruned_step(self, step):
...
         return False
 
+    def rewrite_forward(self):
+        """Rewrite forward to implement block mask operation"""
+        def forward(self, input):
+            block_size = [self.weight.shape[0]//self.block_mask.shape[0], \
+                    self.weight.shape[1]//self.block_mask.shape[1]]
+            mask = self.block_mask.repeat_interleave(block_size[0], dim=0).repeat_interleave(\
+                                                        block_size[1], dim=-1).to(self.weight.device)
+            return F.linear(input, self.weight*mask, self.bias)
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruners.py:294]: layer could be convolution

**Review by wenhuach21 — CHANGES_REQUESTED** [review]

File: neural_compressor/compression/pruner/pruners.py [hunk 7]
```diff
@@ -82,7 +84,8 @@ def get_pruner(config, modules):
     if name in CRITERIA:
         if config["progressive"] == False:
             config['criterion_type'] = name
```
Comment (wenhuach21) [inline neural_compressor/compression/pruner/pruners.py:86]: have you check edthat blockwise criterion only supports NXM

**Review by wenhuach21 — APPROVED** [review]
need to support conv2d later

**Pull request merged**